modules/layers/src/path-layer/tesselator.js

```diff
@@ -0,0 +1,159 @@
+// Copyright (c) 2015 - 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+import {experimental} from '@deck.gl/core';
+const {fillArray} = experimental;
+
+export class TypedArrayManager {
+  constructor({overAlloc = 1} = {}) {
+    this.overAlloc = overAlloc;
+  }
+
+  allocate(Type, size) {
+    if (Type === Uint16Array && size > 65535) {
+      throw new Error('Vertex count exceeds browser index limit');
+    }
+    // TODO - check if available in pool
+    return new Type(size);
+  }
+
+  reallocate(typedArray, newSize, copy = false) {
+    if (newSize <= typedArray.length) {
+      return typedArray;
+    }
+
+    // Allocate at least one element to ensure a valid buffer
+    const allocSize = Math.max(Math.ceil(newSize * this.overAlloc), 1);
+    const newArray = this.allocate(typedArray.constructor, allocSize);
+
+    if (copy) {
+      newArray.set(typedArray);
+    }
+
+    this.release(typedArray);
+    return newArray;
+  }
+
+  release(typedArray) {
+    // TODO - add to pool
+    // logFunctions.onUpdate({
+    //   level: LOG_DETAIL_PRIORITY,
+    //   message: `${attributeName} allocated ${allocCount}`,
+    //   id: this.id
+    // });
+  }
+}
+
+export default class Tesselator {
+  constructor(opts = {}) {
+    const {attributes = {}} = opts;
+
+    this.typedArrayManager = new TypedArrayManager();
+    this.bufferLayout = null;
+    this.instanceCount = 0;
+    this.attributes = {};
+    this._attributeDefs = attributes;
+
+    this.updateGeometry(opts);
+
+    Object.seal(this);
+  }
+
+  /* Public methods */
+  updateGeometry({data, getGeometry, fp64}) {
+    this.data = data;
+    this.getGeometry = getGeometry;
+    this.fp64 = fp64;
+    this._rebuildGeometry();
+  }
+
+  /* Subclass interface */
+
+  // Fill the typed arrays for geometry attributes
+  updateGeometryAttributes() {
```

Comment:
Not sure about this virtual method. If this is going to be a base class for general tesselators, I would imagine that we want to offer the subclass a way to define how to tesselate (fill in a sub-sequence of verts) in each each object's sub-geometry separately.

Reply:
I intentionally kept this vague - it's hard to specify what exactly this method does.

In PathTesselator: fill in positions & deltas attributes https://github.com/uber/deck.gl/pull/2493/files#diff-522e17b773d84f9a994b567c3a9a8183R123

In PolygonTesselator: fill in positions & indices attributes https://github.com/uber/deck.gl/pull/2494/files#diff-2997724296c81145e5e55fdf221b8f7cR100

Reply:
> I intentionally kept this vague - it's hard to specify what exactly this method does...

Well in either case, to support partial updates, this method should have a set of parameters indicating the start and end of the allocated data for that object etc.

To simplify this review: The `Tesselator` base class needs a separate pass. For a separate PR. At least we have a starting point.